### ISSUE TITLE
Allow custom kwargs for forward method of nn.TransformerEncoderLayer, nn.TransformerDecoderLayer.

### ISSUE THREAD

**Issue by kdexd** [pytorch/pytorch#43865]
## 🚀 Feature

A simple feature to allow passing custom keyword arguments for `nn.TransformerEncoderLayer` and `nn.TransformerDecoderLayer`. Ideally, users should be able to flexibly use `custom_encoder` and `custom_decoder` arguments of `nn.Transformer` which can accept arguments of any arbitrary name.

## Motivation

Currently, the `forward method of any custom encoders and decoders require to follow the exact API of `nn.TransformerEncoder` and `nn.TransformerDecoder` — what if custom transformer layers require some extra input tensors? For example, [DETR](https://github.com/facebookresearch/detr) injects sinusoidal positional embedding at every encoder layer. To enable this feature, one needs to reimplement/copy-paste large chunks of the codebase, [like DETR does here](https://github.com/facebookresearch/detr/blob/master/models/transformer.py)

## Pitch

Let the `torch.nn.modules.transformer` Transformer modules follow this common API:

```python
class Transformer(nn.Module):
    def __init__(self, *existing_pytorch_args):
        # ... existing method body.

    def forward(self, *existing_args,  **encoder_kwargs, **decoder_kwargs):
        # Separate encoder_kwargs and decoder_kargs by checking "encoder_" and "decoder_"
        # prefix. We can use `inpect.signature()` for this.
        memory = self.encoder(..., **encoder_kwargs)
        output = self.decoder(..., **decoder_kwargs)
```

Similarly, `nn.TransformerEncoder` should allow `**encoder_layer_kwargs` and `nn.TransformerDecoderLayer` should allow `**decoder_layer_kwargs`.

**Comment by zhangguanheng66** [comment]
I agree this is a good idea. For now, we want to stabilize our APIs for transformer/MHA. We might make the changes if we see consistent requests for this. Thanks @kdexd

**Comment by lukasschmit** [comment]
+1 for this idea. would be really helpful to be able to tunnel into custom transformer layer classes w forward kwargs!
